go/libraries/doltcore/sqle/dtables/diff_iter.go

```diff
@@ -0,0 +1,368 @@
+
+	rows    chan sql.Row
+	errChan chan error
+	cancel  context.CancelFunc
+}
+
+var _ sql.RowIter = prollyDiffIter{}
+
+// newProllyDiffIter produces dolt_diff system table and dolt_diff table
+// function rows. The rows first have the "to" columns on the left and the
+// "from" columns on the right. After the "to" and "from" columns, a commit
+// name, and commit date is also present. The final column is the diff_type
+// column.
+//
+// An example: to_pk, to_col1, to_commit, to_commit_date, from_pk, from_col1, from_commit, from_commit_date, diff_type
+//
+// |targetFromSchema| and |targetToSchema| defines what the schema should be for
+// the row data on the "from" or "to" side. In the above example, both schemas are
+// identical with two columns "pk" and "col1". The dolt diff table function for
+// example can provide two different schemas.
+//
+// The |from| and |to| tables in the DiffPartition may have different schemas
+// than |targetFromSchema| or |targetToSchema|. We convert the rows from the
+// schema of |from| to |targetFromSchema| and the schema of |to| to
+// |targetToSchema|. See the tablediff_prolly package.
+func newProllyDiffIter(ctx *sql.Context, dp DiffPartition, ddb *doltdb.DoltDB, targetFromSchema, targetToSchema schema.Schema) (prollyDiffIter, error) {
+	fromCm := commitInfo2{
+		name: dp.fromName,
+		ts:   (*time.Time)(dp.fromDate),
+	}
+	toCm := commitInfo2{
+		name: dp.toName,
+		ts:   (*time.Time)(dp.toDate),
+	}
+
+	// dp.from may be nil
+	f, fSch, err := tableData(ctx, dp.from, ddb)
+	if err != nil {
+		return prollyDiffIter{}, nil
+	}
+	from := durable.ProllyMapFromIndex(f)
+
+	t, tSch, err := tableData(ctx, dp.to, ddb)
+	if err != nil {
+		return prollyDiffIter{}, nil
+	}
+	to := durable.ProllyMapFromIndex(t)
+
+	fromConverter, err := tablediff_prolly.NewRowConverter(fSch, targetFromSchema)
+	if err != nil {
+		return prollyDiffIter{}, err
+	}
+
+	toConverter, err := tablediff_prolly.NewRowConverter(tSch, targetToSchema)
+	if err != nil {
+		return prollyDiffIter{}, err
+	}
+
+	child, cancel := context.WithCancel(ctx)
+
+	iter := prollyDiffIter{
+		from:          from,
+		to:            to,
+		fromSch:       fSch,
+		toSch:         tSch,
+		targetFromSch: targetFromSchema,
+		targetToSch:   targetToSchema,
+		fromConverter: fromConverter,
+		toConverter:   toConverter,
+		fromCm:        fromCm,
+		toCm:          toCm,
+		rows:          make(chan sql.Row, 64),
+		errChan:       make(chan error),
+		cancel:        cancel,
+	}
+
+	go func() {
+		iter.queueRows(child)
+	}()
+
+	return iter, nil
+}
+
+func (itr prollyDiffIter) Next(ctx *sql.Context) (sql.Row, error) {
+	select {
+	case <-ctx.Done():
+		return nil, ctx.Err()
+	case err := <-itr.errChan:
+		return nil, err
+	case r, ok := <-itr.rows:
+		if !ok {
+			return nil, io.EOF
+		}
+		return r, nil
+	}
+}
+
+func (itr prollyDiffIter) Close(ctx *sql.Context) error {
+	itr.cancel()
+	return nil
+}
+
+func (itr prollyDiffIter) queueRows(ctx context.Context) {
```

Comment:
seems like this doesn't work for keyless tables? we should check the schema and throw an error (for now) when constructing the `prollDiffIter`